tap_mailchimp/streams.py

```diff
@@ -0,0 +1,538 @@
+                                               stream_metadata)
+                singer.write_record(self.stream_name, record)
+                counter.increment()
+            return max_bookmark_field
+
+    def get_bookmark(self, path, default):
+        """Function to get the bookmark at a particular path from the state"""
+        dic = self.state
+        for key in ['bookmarks'] + path:
+            if key in dic:
+                dic = dic[key]
+            else:
+                return default
+        return dic
+
+    def write_bookmark(self, path, value):
+        nested_set(self.state, ['bookmarks'] + path, value)
+        singer.write_state(self.state)
+
+    def format_selected_fields(self):
+        """Given a catalog with selected metadata return a comma separated string
+        of the `data_key.field_name` for every selected field in the catalog plus
+        the `extra_fields` defined below
+
+        The result of this function is expected to be passed to the API as the
+        value of the `fields` parameter
+
+        We add '<data_key>' to the beginning of all the selected fields because
+        fields are nested under the '<data_key>' key in the response object.
+        """
+        mdata = metadata.to_map(self.stream.metadata)
+        fields = self.stream.schema.properties.keys()
+        formatted_field_names = []
+        for field in fields:
+            field_metadata = mdata.get(('properties', field))
+            if should_sync_field(field_metadata.get('inclusion'), field_metadata.get('selected')):
+                formatted_field_names.append(self.data_key+'.'+field)
+
+        default_fields = ['_links', 'total_items', 'constraints', self.data_key+'.'+'_links']
+        formatted_field_names += default_fields
+        if self.extra_fields:
+            formatted_field_names += self.extra_fields
+        return ",".join(formatted_field_names)
+
+    def sync_substream(self, child, parent_record_id):
+        """Function to update the API URL based on parent id and sync child stream"""
+        child_stream_obj = STREAMS.get(child)(self.state, self.client, self.config, \
+            self.catalog, self.selected_stream_names, self.child_streams_to_sync)
+        if child_stream_obj.replication_method == 'INCREMENTAL':
+            # Updated the bookmark path with parent id, as the Tap saves bookmark based on parent's id
+            child_stream_obj.bookmark_path[1] = parent_record_id
+        child_stream_obj.path = self.get_path(parent_record_id, child_stream_obj)
+        child_stream_obj.sync()
+
+    def sync(self, sync_start_date=None):
+        """Function to sync records and call child stream for every records"""
+        last_datetime = sync_start_date
+        max_bookmark_field = last_datetime
+
+        ids = []
+        page_size = int(self.config.get('page_size', DEFAULT_PAGE_SIZE))
+        offset = 0
+        has_more = True
+        while has_more:
+            params = {
```

Comment:
Instead of defining each time, update the offset value in the loop.

Reply:
Updated the code as per suggested.